src/online2bin/online2-tcp-nnet3-decode-faster.cc

```diff
@@ -0,0 +1,451 @@
+      size_t chunk_len = static_cast<size_t>(chunk_length_secs * samp_freq);
+
+      int32 check_update = static_cast<int32>(samp_freq * output_freq);
+      int32 next_check = check_update;
+
+      bool eos = false;
+
+      OnlineIvectorExtractorAdaptationState adaptation_state(
+          feature_info.ivector_extractor_info);
+
+      while (!eos) {
+
+        OnlineNnet2FeaturePipeline feature_pipeline(feature_info);
+        feature_pipeline.SetAdaptationState(adaptation_state);
+
+        OnlineSilenceWeighting silence_weighting(
+            trans_model,
+            feature_info.silence_weighting_config,
+            decodable_opts.frame_subsampling_factor);
+
+        SingleUtteranceNnet3Decoder decoder(decoder_opts, trans_model,
+                                            decodable_info,
+                                            *decode_fst, &feature_pipeline);
+
+        std::vector<std::pair<int32, BaseFloat>> delta_weights;
+
+        while (true) {
+
+          eos = !server.ReadChunk(chunk_len);
+
+          if (!eos) {
+            Vector<BaseFloat> wave_part = server.getChunk();
+            feature_pipeline.AcceptWaveform(samp_freq, wave_part);
+            samp_count += chunk_len;
+
+            if (silence_weighting.Active() &&
+                feature_pipeline.IvectorFeature() != NULL) {
+              silence_weighting.ComputeCurrentTraceback(decoder.Decoder());
+              silence_weighting.GetDeltaWeights(feature_pipeline.NumFramesReady(),
+                                                &delta_weights);
+              feature_pipeline.IvectorFeature()->UpdateFrameWeights(delta_weights);
+            }
+
+            decoder.AdvanceDecoding();
+
+            if (samp_count > next_check) {
+              if (decoder.NumFramesDecoded() > 0) {
+                Lattice lat;
+                decoder.GetBestPath(false, &lat);
+
+                std::string msg = LatticeToString(lat, word_syms);
+
+                server.Write(msg + '\r');
+              }
+
+              next_check += check_update;
+            }
+
+          } else {
+            feature_pipeline.InputFinished();
+
+            decoder.FinalizeDecoding();
+
+            if (decoder.NumFramesDecoded() > 0) {
+              CompactLattice lat;
+              decoder.GetLattice(true, &lat);
+
+              std::string msg = LatticeToString(lat, word_syms);
+
+              server.Write(msg + "\n");
+            } else
+              server.Write("\n");
+
+            server.disconnect();
+            break;
+
+          }
+
+          if (decoder.EndpointDetected(endpoint_opts)) {
+
+            decoder.FinalizeDecoding();
+
+            CompactLattice lat;
+            decoder.GetLattice(true, &lat);
+            std::string msg = LatticeToString(lat, word_syms);
+
+            server.Write(msg + "\n");
+            if (adapt_speaker)
+              feature_pipeline.GetAdaptationState(&adaptation_state);
+            break;
```

Comment:
There is another small issue here regarding the audio samples that were accepted by the feature pipeline but not yet processed into feature frames due to feature pipeline producing feature vectors only after the needed future audio context is available. As it is, the last few samples of audio already accepted by the feature pipeline will effectively be dropped. These dropped samples also have an effect on the first few feature frames extracted for the next segment. So the feature vectors will be extracted as if there was a small gap between the two segments determined by endpointing. The correct thing to do here is to figure out what audio samples are yet to be processed by the feature pipeline and accept those samples after reconstructing the feature pipeline before any new audio samples are accepted. You can take a look at what we have done in [this pykaldi example](https://github.com/pykaldi/pykaldi/blob/master/examples/asr/nnet3-online-recognizer.py#L108-L124) for the exact sequence of things to be done. Then again, this is a rather small issue that probably won't have an appreciable effect on decoding results.

Reply:
Ok, I think I got it. Since we recreate the feature pipeline object for each new segment (which is needed for updating the adaptation state) we loose the last samples from the previous chunk that were still possibly speech.

I don't get one thing though in your code:

` remainder = data[offset:i + chunk_size]`

The offset is calculated purely using the values of the feature pipeline. So it's the offset from the beginning of the last segment, but here you are using it as if it is the offset from the beginning of the file. Shouldn't it be:

` remainder = data[i+offset:i + chunk_size]`

Reply:
I fixed this in the last commit. It's a bit different in my code, since I don't remember the complete audio, but just the last chunk, but the principle is the same.